File: packages/kits/GeniusLocusKit/rust-migrations/tests/fact_extraction_setting_migration_tests.rs
//! Gate tests for the GLK 1.7 → 1.8 fact-extraction-setting migration capsule.
//! Rust twin of Swift `FactExtractionSettingMigrationTests.swift`.
//!
//! Six gate tests (G1–G6): enum default, absent-means-on seeding,
//! explicit-off preservation, stamp advance to V1_8, idempotent re-run,
//! and unrecognised-value fallback.

use std::sync::Arc;

use genius_locus_kit::coordinator::EstateCoordinator;
use genius_locus_kit::estate_preference::{EstatePreferenceKey, EstatePreferenceValue};
use genius_locus_kit::estate_format::{EstateFormatStore, EstateFormatVersion};
use genius_locus_kit::handle::EstateHandle;
use genius_locus_kit_migrations::FactExtractionSettingMigrationExt;
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate_types::OwnerCredentials;
use persistence_kit::Storage;

const NOW: i64 = 1_790_000_000_000; // millis — 2026-09-22

fn make_estate() -> (EstateCoordinator, EstateHandle, Arc<dyn Storage>) {
    let store = Arc::new(InMemoryDrawerStore::new(NOW, None).expect("InMemoryDrawerStore::new"));
    let storage: Arc<dyn Storage> = store
        .storage()
        .expect("InMemoryDrawerStore exposes storage");
    // Stamp V1_7 so the chain considers the estate historical and below current.
    EstateFormatStore::new(Arc::clone(&storage))
        .stamp(EstateFormatVersion::V1_7, NOW)
        .expect("stamp V1_7");
    let mut coord = EstateCoordinator::new();
    let handle = coord
        .open(
            Arc::clone(&store) as Arc<dyn DrawerStore>,
            OwnerCredentials::new("mig18-test-owner"),
            0,
            100,
        )
        .expect("EstateCoordinator::open");
    (coord, handle, storage)
}

// ---------------------------------------------------------------------------
// G1 EstatePreferenceValue default
// ---------------------------------------------------------------------------

#[test]
fn g1_estate_preference_value_default_is_on() {
    let def = EstatePreferenceValue::default();
    assert_eq!(def, EstatePreferenceValue::On);
    assert_eq!(def.as_str(), "on");
    // Roundtrip: from_str returns Some for known values, None for garbage.
    assert_eq!(EstatePreferenceValue::from_str("on"), Some(EstatePreferenceValue::On));
    assert_eq!(EstatePreferenceValue::from_str("off"), Some(EstatePreferenceValue::Off));
    assert_eq!(EstatePreferenceValue::from_str("garbage"), None);
    // The fact-extraction key is stored under the manifest key "fact_extraction".
    assert_eq!(EstatePreferenceKey::FactExtraction.as_str(), "fact_extraction");
    assert_eq!(EstatePreferenceKey::from_str("fact_extraction"), Some(EstatePreferenceKey::FactExtraction));
}

// ---------------------------------------------------------------------------
// G2 absent-means-on seeding
// ---------------------------------------------------------------------------

#[test]
fn g2_migration_seeds_fact_extraction_on_when_absent() {
    let (coord, handle, storage) = make_estate();

    // Key must be absent before the capsule runs.
    let raw_before = coord
        .estate_for(&handle)
        .ok()
        .and_then(|e| e.meta(EstatePreferenceKey::FactExtraction.as_str()).ok())
        .flatten();
    assert!(raw_before.is_none(), "key must be absent before migration");

    // The public accessor returns On even when absent (absent-means-on).
    let before = coord
        .provisioned_preference(&handle, EstatePreferenceKey::FactExtraction)
        .expect("provisioned_preference");
    assert_eq!(before, EstatePreferenceValue::On);

    // Run the capsule.
    coord
        .run_fact_extraction_setting_migration(&handle, NOW)
        .expect("capsule succeeded");

    // The key must now be physically present with value "on".
    let raw_after = coord
        .estate_for(&handle)
        .ok()
        .and_then(|e| e.meta(EstatePreferenceKey::FactExtraction.as_str()).ok())
        .flatten();
    assert_eq!(raw_after.as_deref(), Some("on"), "migration seeded fact_extraction = on");

    let after = coord
        .provisioned_preference(&handle, EstatePreferenceKey::FactExtraction)
        .expect("provisioned_preference after migration");
    assert_eq!(after, EstatePreferenceValue::On);
    drop(storage);
}

// ---------------------------------------------------------------------------
// G3 explicit-off preserved
// ---------------------------------------------------------------------------

#[test]
fn g3_migration_preserves_explicit_off() {
    let (coord, handle, storage) = make_estate();

    // Pre-write Off through the public provisioner before migration.
    coord
        .provision_preference(&handle, EstatePreferenceKey::FactExtraction, EstatePreferenceValue::Off)
        .expect("provision Off");

    // Run the capsule.
    coord
        .run_fact_extraction_setting_migration(&handle, NOW)
        .expect("capsule succeeded");

    // The capsule must not overwrite an existing value.
    let after = coord
        .provisioned_preference(&handle, EstatePreferenceKey::FactExtraction)
        .expect("provisioned after migration");
    assert_eq!(after, EstatePreferenceValue::Off, "capsule must not overwrite Off");

    let raw = coord
        .estate_for(&handle)
        .ok()
        .and_then(|e| e.meta(EstatePreferenceKey::FactExtraction.as_str()).ok())
        .flatten();
    assert_eq!(raw.as_deref(), Some("off"));
    // The stamp advances to V1_8 even when the value was pre-set.
    let stamp = EstateFormatStore::new(Arc::clone(&storage))
        .read_if_present()
        .expect("read_if_present");
    assert_eq!(stamp, Some(EstateFormatVersion::V1_8), "capsule stamps V1_8 even when value is pre-set");
}

// ---------------------------------------------------------------------------
// G4 stamp advance to V1_8
// ---------------------------------------------------------------------------

#[test]
fn g4_migration_stamps_v1_8() {
    let (coord, handle, storage) = make_estate();

    coord
        .run_fact_extraction_setting_migration(&handle, NOW)
        .expect("capsule succeeded");

    let stamp = EstateFormatStore::new(Arc::clone(&storage))
        .read_if_present()
        .expect("read_if_present");
    assert_eq!(stamp, Some(EstateFormatVersion::V1_8), "capsule must stamp V1_8");
    // The capsule stamps V1_8 in isolation; later capsules carry the estate
    // on to `EstateFormatVersion::CURRENT`, which sits at or above V1_8.
    assert!(EstateFormatVersion::V1_8 <= EstateFormatVersion::CURRENT);
}

// ---------------------------------------------------------------------------
// G5 idempotent re-run
// ---------------------------------------------------------------------------

#[test]
fn g5_migration_is_idempotent() {
    let (coord, handle, storage) = make_estate();

    // First run: seeds and stamps.
    coord
        .run_fact_extraction_setting_migration(&handle, NOW)
        .expect("first run succeeded");
    // Second run: must not fail and must leave the value unchanged.
    coord
        .run_fact_extraction_setting_migration(&handle, NOW)
        .expect("second run succeeded");

    let after = coord
        .provisioned_preference(&handle, EstatePreferenceKey::FactExtraction)
        .expect("provisioned after second run");
    assert_eq!(after, EstatePreferenceValue::On, "second run must leave value at On");

    let stamp = EstateFormatStore::new(Arc::clone(&storage))
        .read_if_present()
        .expect("read_if_present");
    assert_eq!(stamp, Some(EstateFormatVersion::V1_8), "second run must leave stamp at V1_8");
}

// ---------------------------------------------------------------------------
// G6 accessor unrecognised-value fallback
// ---------------------------------------------------------------------------

#[test]
fn g6_provisioned_preference_returns_on_for_unrecognised_value_garbage() {
    let (coord, handle, _storage) = make_estate();

    // Write "garbage" directly via set_meta, bypassing the typed provisioner,
    // so the accessor's unrecognised-value branch (from_str returns None →
    // unwrap_or_default → On) actually executes.
    coord
        .estate_for(&handle)
        .expect("estate open")
        .set_meta(EstatePreferenceKey::FactExtraction.as_str(), "garbage")
        .expect("set_meta succeeded");

    // The accessor must degrade to On — the fail-quiet fallback.
    let result = coord
        .provisioned_preference(&handle, EstatePreferenceKey::FactExtraction)
        .expect("provisioned_preference");
    assert_eq!(result, EstatePreferenceValue::On, "unrecognised value must fall back to On");
}

File: packages/kits/GeniusLocusKit/Tests/GLKMigrationV1_7ToV1_8Tests/FactExtractionSettingMigrationTests.swift
// Tests for the GLK 1.7 → 1.8 migration capsule and the EstatePreference
// accessor pair on the fact_extraction key. Seven tests: the six capsule gates G1–G6 (enum default,
// absent-means-on seeding, explicit-off preservation, stamp advance,
// idempotent re-run, unrecognised-value fallback) and one chain test (an
// estate stamped v1_7 driven through GLKMigrationCatalog.prepare reaches
// .current with the key seeded).
//
// The catalog's 1.6→1.7 vacuum guard (`found < .v1_7`) is gated where the
// vacuum fixture lives: WholeRecordFloatVacuumMigrationTests §8 seeds the
// rows the vacuum would delete and asserts they survive a chain from v1_7.
// This target carries no SynapseKit or CorpusKit product and cannot seed
// those rows.
//
// Enabled by the MigrationV1_7ToV1_8 trait (GLK_MIGRATION_V1_7_TO_V1_8 define).

#if GLK_MIGRATION_V1_7_TO_V1_8

import Foundation
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import PersistenceKitSQLite
import Testing
import GeniusLocusKitMigrations
@testable import GeniusLocusKit
@testable import GLKMigrationV1_7ToV1_8

private let testOwner = OwnerCredentials(ownerIdentifier: "test-owner-mig18")
private let testNow = Date(timeIntervalSince1970: 1_790_000_000)

private func makeEstate(
    stampedAt stamp: EstateFormatVersion? = .v1_7
) async throws -> (kit: GeniusLocusKit, handle: EstateHandle, storage: any Storage) {
    let storage = InMemoryStorage(
        configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
    _ = try await LocusKit.Estate.create(storage: storage, owner: testOwner)
    if let stamp {
        try await EstateFormatStore(storage: storage).stamp(stamp, now: testNow)
    }
    let kit = GeniusLocusKit()
    let handle = try await kit.open(
        storage: storage, owner: testOwner, identityKeyStore: InMemoryEstateIdentityKeyStore())
    return (kit, handle, storage)
}

// MARK: - G1 EstatePreferenceKey defaultValue

@Test("G1: EstatePreferenceKey.defaultValue returns the correct per-key default")
func estatePreferenceKeyDefaultValue() {
    // The six on/off switches default to .on; the extractor key defaults to .nuextract.
    #expect(EstatePreferenceKey.factExtraction.defaultValue == .on)
    #expect(EstatePreferenceKey.factExtraction.defaultValue.rawValue == "on")
    #expect(EstatePreferenceKey.factExtractor.defaultValue == .nuextract)
    #expect(EstatePreferenceKey.factExtraction.rawValue == "fact_extraction")
    #expect(EstatePreferenceKey.factExtractor.rawValue == "fact_extractor")
    // Roundtrip the rawValue initializer for on/off.
    #expect(EstatePreferenceValue(rawValue: "on") == .on)
    #expect(EstatePreferenceValue(rawValue: "off") == .off)
    // Roundtrip for the extractor values.
    #expect(EstatePreferenceValue(rawValue: "nuextract") == .nuextract)
    #expect(EstatePreferenceValue(rawValue: "apple") == .apple)
    #expect(EstatePreferenceValue(rawValue: "garbage") == nil)
}

// MARK: - G2 absent-means-on seeding

@Test("G2: migration seeds fact_extraction=on when the key is absent")
func migrationSeedsFactExtractionWhenAbsent() async throws {
    let (kit, handle, storage) = try await makeEstate()

    // The key must be absent before migration runs.
    let rawBefore = try? await kit.estate(for: handle).meta(key: EstatePreferenceKey.factExtraction.rawValue)
    #expect(rawBefore == nil, "key must be absent before migration")

    // The public accessor returns .on even when absent (absent-means-on).
    let settingBefore = try await kit.provisionedPreference(.factExtraction, for: handle)
    #expect(settingBefore == .on)

    // Run the capsule.
    try await kit.runFactExtractionSettingMigration(handle: handle, now: testNow)

    // The key must now be physically present and set to "on".
    let rawAfter = try? await kit.estate(for: handle).meta(key: EstatePreferenceKey.factExtraction.rawValue)
    #expect(rawAfter == "on", "migration must seed fact_extraction = on")
    let settingAfter = try await kit.provisionedPreference(.factExtraction, for: handle)
    #expect(settingAfter == .on)
}

// MARK: - G3 explicit-off preserved

@Test("G3: migration preserves an explicit .off setting")
func migrationPreservesExplicitOff() async throws {
    let (kit, handle, storage) = try await makeEstate()

    // Pre-write .off through the provisioner before migration.
    try await kit.provisionPreference(.factExtraction, .off, for: handle)

    // Run the capsule.
    try await kit.runFactExtractionSettingMigration(handle: handle, now: testNow)

    // The capsule must not overwrite an existing value; the stamp advances to v1_8.
    let after = try await kit.provisionedPreference(.factExtraction, for: handle)
    #expect(after == .off, "capsule must not overwrite an existing value")
    let raw = try? await kit.estate(for: handle).meta(key: EstatePreferenceKey.factExtraction.rawValue)
    #expect(raw == "off")
    let stamp = try await EstateFormatStore(storage: storage).readIfPresent()
    #expect(stamp == .v1_8, "capsule stamps v1_8 even when the value is pre-set")
}

// MARK: - G4 stamp advance to v1_8

@Test("G4: migration stamps the estate format to v1_8")
func migrationStampsV1_8() async throws {
    let (kit, handle, storage) = try await makeEstate()

    try await kit.runFactExtractionSettingMigration(handle: handle, now: testNow)

    let stamp = try await EstateFormatStore(storage: storage).readIfPresent()
    #expect(stamp == .v1_8, "capsule must stamp v1_8")
}

// MARK: - G5 idempotent re-run

@Test("G5: running the capsule twice is idempotent")
func migrationIsIdempotent() async throws {
    let (kit, handle, storage) = try await makeEstate()

    // First run: seeds and stamps.
    try await kit.runFactExtractionSettingMigration(handle: handle, now: testNow)
    // Second run: must not throw.
    try await kit.runFactExtractionSettingMigration(handle: handle, now: testNow)

    let setting = try await kit.provisionedPreference(.factExtraction, for: handle)
    #expect(setting == .on, "second run must leave value at .on")
    let stamp = try await EstateFormatStore(storage: storage).readIfPresent()
    #expect(stamp == .v1_8, "second run must leave stamp at v1_8")
}

// MARK: - G6 accessor unrecognised-value fallback

@Test("G6: provisionedPreference(.factExtraction) returns .on when stored value is unrecognised (\"garbage\")")
func provisionedPreferenceReturnsOnForUnrecognisedValue_garbage() async throws {
    let (kit, handle, _) = try await makeEstate()

    // Write an unrecognised string directly via setMeta, bypassing the
    // typed provisioner, so the accessor's unrecognised-value branch runs.
    try await kit.estate(for: handle).setMeta(
        key: EstatePreferenceKey.factExtraction.rawValue,
        value: "garbage")

    // The accessor must degrade to .on — the fail-quiet fallback.
    let result = try await kit.provisionedPreference(.factExtraction, for: handle)
    #expect(result == .on, "unrecognised value must fall back to .on")
}

// MARK: - Chain from v1_7 reaches current through the 1.7→1.8 capsule

// An estate stamped v1_7 driven through the chain entry point ends at .current
// with `fact_extraction` seeded: the seeded key is what shows the chain invoked
// this capsule rather than stamping on its own. Whether the chain skipped the
// 1.6→1.7 vacuum on the way is not observable on this bare fixture; that guard
// is gated by WholeRecordFloatVacuumMigrationTests §8.
@Test("Chain from v1_7: reaches current with fact_extraction seeded")
func chainFromV1_7EstateReachesCurrentFormat() async throws {
    let (kit, handle, storage) = try await makeEstate(stampedAt: .v1_7)
    let rawBefore = try? await kit.estate(for: handle).meta(key: EstatePreferenceKey.factExtraction.rawValue)
    #expect(rawBefore == nil, "key must be absent before the chain runs")
    let prep = try await GLKMigrationCatalog.prepare(kit: kit, handle: handle, now: testNow, offlineUpgrade: true)
    #expect(prep.format == .current)
    #expect(try await EstateFormatStore(storage: storage).readIfPresent() == .current)
    let rawAfter = try? await kit.estate(for: handle).meta(key: EstatePreferenceKey.factExtraction.rawValue)
    #expect(rawAfter == "on", "the chain must run the 1.7→1.8 capsule, which seeds the key")
}

#endif
